docs/framework/wcf/samples/custom-lifetime.md

```diff
@@ -1,219 +1,227 @@
-  
-#### To set up, build, and run the sample  
-  
-1.  Ensure that you have performed the [One-Time Setup Procedure for the Windows Communication Foundation Samples](../../../../docs/framework/wcf/samples/one-time-setup-procedure-for-the-wcf-samples.md).  
-  
-2.  To build the C# or Visual Basic .NET edition of the solution, follow the instructions in [Building the Windows Communication Foundation Samples](../../../../docs/framework/wcf/samples/building-the-samples.md).  
-  
-3.  To run the sample in a single- or cross-machine configuration, follow the instructions in [Running the Windows Communication Foundation Samples](../../../../docs/framework/wcf/samples/running-the-samples.md).  
-  
+> Using thread locking is not a recommended approach because serialization can severely affect the performance of your application.
+
+A private member variable is used in the `CustomLifetimeLease` class to track the `IsIdle` value. Each time the value of <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider> is retrieved, the `IsIdle` private member is returned and reset to `false`. It is essential to set this value to `false` in order to make sure the Dispatcher calls the <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider.NotifyIdle%2A> method.
+
+```csharp
+public bool IsIdle(InstanceContext instanceContext)
+{
+    get
+    {
+        lock (thisLock)
+        {
+            //...
+            bool idleCopy = isIdle;
+            isIdle = false;
+            return idleCopy;
+        }
+    }
+}
+```
+
+If the `IInstanceContextProvider.IsIdle` property returns `false`, the Dispatcher registers a callback function by using the `NotifyIdle` method. This method receives a reference to the <xref:System.ServiceModel.InstanceContext> being released. Therefore, the sample code can query the `ICustomLease` type extension and check the `ICustomLease.IsIdle` property in the extended state.
+
+```csharp
+public void NotifyIdle(InstanceContextIdleCallback callback,
+            InstanceContext instanceContext)
+{
+    lock (thisLock)
+    {
+       ICustomLease customLease =
+           instanceContext.Extensions.Find<ICustomLease>();
+       customLease.Callback = callback;
+       isIdle = customLease.IsIdle;
+       if (isIdle)
+       {
+             callback(instanceContext);
+       }
+    }
+}
+```
+
+Before the `ICustomLease.IsIdle` property is checked, the Callback property needs to be set as this is essential for `CustomLeaseExtension` to notify the Dispatcher when it becomes idle. If `ICustomLease.IsIdle` returns `true`, the `isIdle` private member is simply set in `CustomLifetimeLease` to `true` and calls the callback method. Because the code holds a lock, other threads cannot change the value of this private member. And the next time Dispatcher checks the `IInstanceContextProvider.IsIdle`, it returns `true` and lets Dispatcher release the instance.
```

Comment:
And the next time the Dispatcher calls the <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider.IsIdle> method,....